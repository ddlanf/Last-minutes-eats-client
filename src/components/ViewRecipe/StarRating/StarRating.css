.star-rating{
    display: inline;
    font-size: 9vw;
    z-index: 7;
}

.view-recipe-rate{
    display: block;
    font-size: 5vw;
}

.view-recipe-stars{
    width: 100%;
    display: block;
}

@media screen and (min-width: 1000px) {

    .view-recipe-rate{
        display: block;
        font-size: 3vw;
    }

    .star-rating{
        display: inline;
        font-size: 3.7vw;
    }
}