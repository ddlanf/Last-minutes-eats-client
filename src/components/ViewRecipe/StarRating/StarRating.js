import React, { Component } from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import RatingApiService from '../../../services/ratings-api-service'
import TokenService from '../../../services/token-service'
import './StarRating.css'

export default class StarRating extends Component {

    
    constructor(props){
       
       super(props)
       const { recipeId } = this.props
       const ratingToken = TokenService.getRatingToken(recipeId)

       this.state = {
            stars: !!ratingToken ?  this.returnStarArray(parseInt(ratingToken)) : [false, false, false, false, false],
            disabled: !!ratingToken
        }
    }

    returnStarArray = (rating) =>{
        
        const newState = []

        for(let i = 1; i < 6; i++){
            if(rating >= i){ newState.push(true) }
            else{ newState.push(false) }
        }

        return newState
    }

    showSolid = (index) =>{
        if(!this.state.disabled){
            const newState = this.state.stars
            for(let i = 0; i < index + 1; i++){
                newState[i] = true
            }
            
            this.setState({ stars : newState })
        }
    }

    showRegular = (index) =>{
        if(!this.state.disabled){
            const newState = []

            for(let i = 0; i < 5; i++){
                if(index => i){ newState[i] = false }
            }
            this.setState({ stars : newState })
        }
    }

    submitRating = (rating) =>{
        
        if(!this.state.disabled){
            const newRating = { rating }
            const newState = this.state.stars
            for(let i = 0; i < rating; i++){
                newState[i] = true
            }
            this.setState({ stars : newState, disabled: true })
            
            const { recipeId } = this.props

            RatingApiService.postRating(newRating, recipeId)
            .then(newRating => { 
                TokenService.saveRatingToken(recipeId, newRating.rating) 
            })
            .catch(res =>{ this.setState({ error: res.error})})
        }
    }

    renderStarRating(){
        return [1,2,3,4,5].map((num, index)=>{
            return (
                <div 
                    key={num} 
                    className="star-rating">
                     <FontAwesomeIcon  
                        style={{display : (!this.state.stars[index]) ? "inline" : 'none'}} 
                        onMouseEnter={() => this.showSolid(index)} className="view-recipe-star" 
                        icon={['far', 'star']}/>
                    <FontAwesomeIcon  
                        onClick={()=> this.submitRating(num)} 
                        style={{display : (this.state.stars[index]) ? "inline" : 'none'}} 
                        onMouseLeave={() => this.showRegular(num)} 
                        className="view-recipe-star" 
                        icon={['fas', 'star']}/>
                </div>
            )
        })
    }

    render() {
        return (
            <>
              <label className="view-recipe-rate">
                    {!this.disabled ? "Your rating" : "Rate this recipe"}
              </label>
              <span className="view-recipe-stars">
                    {this.renderStarRating()}
              </span>
            </>
        )
    }
}
