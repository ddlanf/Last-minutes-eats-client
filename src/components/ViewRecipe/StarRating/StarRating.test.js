import React from 'react';
import ReactDOM from 'react-dom';
import StarRating from './StarRating';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom'

describe('Make component', () => {
    

    it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<BrowserRouter>
                        <StarRating/>
                    </BrowserRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
    });

    it('renders the UI as expected', () => {
        const tree = renderer
          .create(
                <BrowserRouter>
                    <StarRating/>
                </BrowserRouter>)
          .toJSON();
        expect(tree).toMatchSnapshot();  
      });
});