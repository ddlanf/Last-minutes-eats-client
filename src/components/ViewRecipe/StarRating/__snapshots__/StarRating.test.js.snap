// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Make component renders the UI as expected 1`] = `
Array [
  <label
    className="view-recipe-rate"
  >
    Your rating
  </label>,
  <span
    className="view-recipe-stars"
  >
    <div
      className="star-rating"
    >
      <i
        className="fa"
      />
      <i
        className="fa"
      />
    </div>
    <div
      className="star-rating"
    >
      <i
        className="fa"
      />
      <i
        className="fa"
      />
    </div>
    <div
      className="star-rating"
    >
      <i
        className="fa"
      />
      <i
        className="fa"
      />
    </div>
    <div
      className="star-rating"
    >
      <i
        className="fa"
      />
      <i
        className="fa"
      />
    </div>
    <div
      className="star-rating"
    >
      <i
        className="fa"
      />
      <i
        className="fa"
      />
    </div>
  </span>,
]
`;
